File: components/amp-utility/python/I2C.py
# * coding: UTF8 *
"""

=================================================================================================

"""

def open(type):
   """
   打开并根据board.json配置实例

   :param type: 对象类型
   :returns: 0: 成功，其他: 失败
   :raises OSError: EINVAL
   """
   pass

def write(dataBuffer):
   """
   发送数据，该函数为阻塞函数，I2C发送完成后才会返

   :param dataBuffer: 待写入的数据
   :returns: 0: 成功，其他: 失败
   :raises OSError: EINVAL
   """
   pass

def read(dataBuffer):
   """
   主动读取指定bytes的I2C数据。该函数为阻塞函数，会一直等待读取指定bytes的数据

   :param dataBuffer: 读出来数据的存储空间
   :returns: >0: 成功读取数据的字节数，其他: 失败
   :raises OSError: EINVAL
   """
   pass

def writeReg(addr, dataBuffer):
   """
   向寄存器地址memaddr发送数据，该函数为阻塞函数，I2C发送完成后才会返回

   :param addr: 设备地址
   :param dataBuffer: 待写入的数据
   :returns: 0: 成功，其他: 失败
   :raises OSError: EINVAL
   """
   pass

def readReg(addr, dataBuffer):
   """

    主动向寄存器地址memaddr读取指定bytes的I2C数据。该函数为阻塞函数，会一直等待读取指定bytes的数据
   
   :param addr: 设备地址
   :param dataBuffer: 读出来数据的存储空间
   :returns: >0: 成功读取数据的字节数，其他: 失败
   :raises OSError: EINVAL
   """
   pass

def close():
   """
   关闭实例

   :param 空:
   :returns: 0: 成功，其他: 失败
   :raises OSError: EINVAL
   """
   pass

File: components/amp/jslib/src/i2c.js
import * as I2C from 'I2C'

function byteArrayToArrayBuffer(byteArray) {
    return new Uint8Array(byteArray).buffer
}
class HW_I2C {
    constructor(options) {
        if (!options || !options.id) {
            throw new Error("options is invalid");
        }

        this.options = {
            id: options.id
        };

        this.success = options.success || function(){};
        this.fail = options.fail || function(){};
        this._open();
    }

    _open() {
        this.i2cInstance = I2C.open(this.options.id);
        if (!this.i2cInstance) {
            this.fail();
            return;
        }
        this.success();
    }

    write(data) {
        if (!this.i2cInstance || !data) {
            throw new Error("i2c not init or params is invalid");
        }
        this.i2cInstance.write(byteArrayToArrayBuffer(data));
    }

    read(bytes) {
        if (!this.i2cInstance || !bytes) {
            throw new Error("i2c not init or params is invalid");
        }
        return Array.prototype.slice.call(new Uint8Array(this.i2cInstance.read(bytes)));
    };

    writeMem(memaddr, data) {
        if (!this.i2cInstance) {
            throw new Error("i2c not init or params is invalid");
        }
        this.i2cInstance.writeReg(memaddr, byteArrayToArrayBuffer(data));
    }

    readMem(memaddr, bytes) {
        if (!this.i2cInstance) {
            throw new Error("i2c not init or params is invalid");
        }
        return Array.prototype.slice.call(new Uint8Array(this.i2cInstance.readReg(memaddr, bytes)));
    };

    close() {
        if (!this.i2cInstance) {
            throw new Error("i2c not init");
        }
        this.i2cInstance.close();
    };
}

export function open(options) {
    return new HW_I2C(options);
}
